File: apps/mac/Sources/Features/Work/WorkSearchText.swift
// SPDX-License-Identifier: LicenseRef-tokenstat-source-available
import Foundation
import Darwin

/// Literal search terms, never regular expressions or database syntax.
struct WorkSearchQuery: Equatable, Sendable {
    static let maximumCharacters = 512
    static let maximumUTF8Bytes = 16 * 1024
    let terms: [String]

    enum Invalid: Error { case tooLong }

    init(_ text: String) throws {
        guard text.utf8.count <= Self.maximumUTF8Bytes,
              text.count <= Self.maximumCharacters else { throw Invalid.tooLong }
        var seen = Set<String>()
        terms = text.split(whereSeparator: \.isWhitespace).map {
            WorkSearchText.normalize(String($0))
        }.filter { !$0.isEmpty && seen.insert($0).inserted }
    }
}

/// Disposable memory only. Character boundaries map folded matches back to
/// the original words so accents and multi-scalar emoji remain intact.
struct WorkSearchText: Sendable {
    struct Excerpt: Equatable, Sendable {
        let text: String
        /// UTF-16 ranges in text, suitable for native attributed strings.
        let highlights: [NSRange]
    }

    static let maximumExcerptCharacters = 240
    let original: String
    private let foldedUTF8: [UInt8]
    private let signature: [UInt64]?

    struct Needle: Sendable {
        let bytes: [UInt8]
        fileprivate let signature: [UInt64]?
        init(_ text: String) {
            bytes = Array(text.utf8)
            signature = WorkSearchText.signature(bytes)
        }
    }

    /// A small rejection filter for long messages. Every consecutive byte
    /// triple sets a bit; collisions only cause a full literal search, never
    /// a missing match. Short labels do not need the extra storage.
    private static func signature(_ bytes: [UInt8]) -> [UInt64]? {
        guard bytes.count >= 3 else { return nil }
        var bits = [UInt64](repeating: 0, count: 16)
        for i in 0..<(bytes.count - 2) {
            let triple = UInt32(bytes[i]) << 16 | UInt32(bytes[i + 1]) << 8 | UInt32(bytes[i + 2])
            let mixed = triple &* 0x9e3779b1
            let bit = Int((mixed ^ (mixed >> 16)) & 1023)
            bits[bit >> 6] |= UInt64(1) << (bit & 63)
        }
        return bits
    }

    init(_ text: String) {
        original = text
        foldedUTF8 = Array(Self.normalize(text).utf8)
        signature = foldedUTF8.count >= 128 ? Self.signature(foldedUTF8) : nil
    }

    static func normalize(_ text: String) -> String {
        text.folding(options: .caseInsensitive, locale: Locale(identifier: "en_US_POSIX"))
            .precomposedStringWithCanonicalMapping
    }

    func contains(_ term: String) -> Bool { contains(Needle(term)) }

    /// Both sides are already case-folded NFC. Literal UTF-8 byte matching
    /// has the same boundaries as a valid Unicode needle, without rebuilding
    /// Foundation string-search state for every field of every message.
    func contains(_ term: Needle) -> Bool {
        guard !term.bytes.isEmpty else { return false }
        if let signature, let required = term.signature {
            for index in signature.indices where signature[index] & required[index] != required[index] { return false }
        }
        return foldedUTF8.withUnsafeBufferPointer { haystack in
            term.bytes.withUnsafeBufferPointer { needle in
                guard let source = haystack.baseAddress, let target = needle.baseAddress else { return false }
                return memmem(source, haystack.count, target, needle.count) != nil
            }
        }
    }

    func excerpt(for query: WorkSearchQuery) -> Excerpt {
        let folded = String(decoding: foldedUTF8, as: UTF8.self)
        // Mapping costs more memory than the text. Build it only for visible
        // result excerpts, not for every indexed message.
        let rawMatches = query.terms.compactMap { term in
            folded.range(of: term, options: .literal).map { NSRange($0, in: folded) }
        }.sorted { $0.location < $1.location }
        let firstOffset = rawMatches.first?.location ?? 0
        let characters = Array(original)
        var foldedOffsets = [0]
        var offset = 0
        var contextRemaining = Self.maximumExcerptCharacters
        for character in characters {
            if character.asciiValue != nil {
                // Swift treats CRLF as one Character with ASCII value LF,
                // but its normalized UTF-16 position still advances by two.
                offset += character == "\r\n" ? 2 : 1
            } else { offset += Self.normalize(String(character)).utf16.count }
            foldedOffsets.append(offset)
            if offset >= firstOffset {
                contextRemaining -= 1
                if contextRemaining == 0 { break }
            }
        }
        func boundary(at offset: Int, roundingUp: Bool) -> Int {
            var low = 0
            var high = foldedOffsets.count
            while low < high {
                let middle = (low + high) / 2
                if foldedOffsets[middle] < offset { low = middle + 1 } else { high = middle }
            }
            if low < foldedOffsets.count, foldedOffsets[low] == offset { return low }
            return roundingUp ? min(low, foldedOffsets.count - 1) : max(0, low - 1)
        }

        let matches = rawMatches.compactMap { utf16 -> Range<Int>? in
            guard utf16.location < offset else { return nil }
            // A fold can expand one character (ß to ss). Highlight the
            // whole original character, including a match within an expansion.
            return boundary(at: utf16.location, roundingUp: false)..<boundary(at: NSMaxRange(utf16), roundingUp: true)
        }

        let budget = Self.maximumExcerptCharacters
        let first = matches.first?.lowerBound ?? 0
        var start = max(0, first - 48)
        let earliest = max(0, start - 24)
        while start > earliest, !characters[start - 1].isWhitespace { start -= 1 }
        // A long unbroken token has no useful word boundary. Keep the
        // original context budget rather than chasing its beginning.
        if start == earliest, start > 0, !characters[start - 1].isWhitespace { start = max(0, first - 48) }
        let prefix = start > 0 ? 1 : 0
        var end = min(characters.count, start + budget - prefix)
        if end < characters.count { end -= 1 }
        var bytes = prefix * "…".utf8.count
        for index in start..<end {
            let length = String(characters[index]).utf8.count
            if bytes + length + "…".utf8.count > 16 * 1024 { end = index; break }
            bytes += length
        }
        let suffix = end < characters.count
        let body = String(characters[start..<end])
        let text = (prefix == 1 ? "…" : "") + body + (suffix ? "…" : "")
        var visibleMatches = matches.compactMap { match -> Range<Int>? in
            let lower = max(start, match.lowerBound)
            let upper = min(end, match.upperBound)
            return lower < upper ? lower..<upper : nil
        }
        let searchEnd = String.Index(utf16Offset: foldedOffsets[end], in: folded)
        for term in query.terms {
            var searchStart = String.Index(utf16Offset: foldedOffsets[start], in: folded)
            while searchStart < searchEnd,
                  let range = folded.range(of: term, options: .literal, range: searchStart..<searchEnd) {
                let utf16 = NSRange(range, in: folded)
                visibleMatches.append(boundary(at: utf16.location, roundingUp: false)..<boundary(at: NSMaxRange(utf16), roundingUp: true))
                searchStart = range.upperBound
            }
        }
        var ranges: [Range<Int>] = []
        for clipped in visibleMatches.sorted(by: { $0.lowerBound < $1.lowerBound }) {
            if let previous = ranges.last, previous.upperBound >= clipped.lowerBound {
                ranges[ranges.count - 1] = previous.lowerBound..<max(previous.upperBound, clipped.upperBound)
            } else {
                ranges.append(clipped)
            }
        }
        let highlights = ranges.map { range in
            NSRange(location: prefix + String(characters[start..<range.lowerBound]).utf16.count,
                    length: String(characters[range]).utf16.count)
        }
        return Excerpt(text: text, highlights: highlights)
    }

}

File: crates/tokenstat-host/src/work_search_text.rs
// SPDX-License-Identifier: LicenseRef-tokenstat-source-available
//! Canonical Unicode search keys and original-text excerpts. Folded text is
//! for matching only, never shown or used as an accessibility label.

use unicase::UniCase;
use unicode_normalization::UnicodeNormalization;
use unicode_segmentation::UnicodeSegmentation;

use crate::work_search::{Highlight, MAX_EXCERPT_CHARACTERS};

pub fn normalize(text: &str) -> String {
    let decomposed: String = text.nfd().collect();
    UniCase::new(decomposed).to_folded_case().nfc().collect()
}

pub fn terms(query: &str) -> Vec<String> {
    let mut result = Vec::new();
    for word in query.split_whitespace() {
        let word = normalize(word);
        if !word.is_empty() && !result.contains(&word) {
            result.push(word);
        }
    }
    result
}

pub fn excerpt(original: &str, terms: &[String]) -> (String, Vec<Highlight>) {
    let graphemes: Vec<&str> = original.graphemes(true).collect();
    let mut folded = String::new();
    let mut offsets = vec![0];
    for grapheme in &graphemes {
        folded.push_str(&normalize(grapheme));
        offsets.push(folded.len());
    }
    let matches: Vec<_> = terms
        .iter()
        .filter(|term| !term.is_empty())
        .filter_map(|term| {
            folded.find(term).map(|start| {
                let end = start + term.len();
                let first = offsets
                    .partition_point(|value| *value <= start)
                    .saturating_sub(1);
                let last = offsets
                    .partition_point(|value| *value < end)
                    .min(graphemes.len());
                first..last
            })
        })
        .collect();
    let first = matches.iter().map(|range| range.start).min().unwrap_or(0);
    let mut start = first.saturating_sub(48);
    let earliest = start.saturating_sub(24);
    while start > earliest && !graphemes[start - 1].chars().all(char::is_whitespace) {
        start -= 1;
    }
    if start == earliest && start > 0 && !graphemes[start - 1].chars().all(char::is_whitespace) {
        start = first.saturating_sub(48);
    }
    let prefix = usize::from(start > 0);
    let mut end = graphemes.len().min(start + MAX_EXCERPT_CHARACTERS - prefix);
    if end < graphemes.len() {
        end -= 1;
    }
    // A single grapheme may contain arbitrarily many combining marks.
    // Never split it, and never let it create an unbounded wire excerpt.
    let mut bytes = prefix * "…".len();
    for (index, grapheme) in graphemes.iter().enumerate().take(end).skip(start) {
        if bytes + grapheme.len() + "…".len() > 16 * 1024 {
            end = index;
            break;
        }
        bytes += grapheme.len();
    }
    let mut text = if prefix == 1 {
        "…".to_string()
    } else {
        String::new()
    };
    text.push_str(&graphemes[start..end].concat());
    if end < graphemes.len() {
        text.push('…');
    }
    let mut ranges: Vec<_> = matches
        .into_iter()
        .filter_map(|range| {
            let lower = range.start.max(start);
            let upper = range.end.min(end);
            (lower < upper).then_some(lower..upper)
        })
        .collect();
    let visible = &folded[offsets[start]..offsets[end]];
    for term in terms.iter().filter(|term| !term.is_empty()) {
        for (position, _) in visible.match_indices(term.as_str()) {
            let lower = offsets
                .partition_point(|offset| *offset <= offsets[start] + position)
                .saturating_sub(1);
            let upper = offsets
                .partition_point(|offset| *offset < offsets[start] + position + term.len())
                .min(end);
            ranges.push(lower..upper);
        }
    }
    ranges.sort_by_key(|range| range.start);
    let mut merged: Vec<std::ops::Range<usize>> = Vec::new();
    for range in ranges {
        if let Some(previous) = merged
            .last_mut()
            .filter(|previous| previous.end >= range.start)
        {
            previous.end = previous.end.max(range.end);
        } else {
            merged.push(range);
        }
    }
    let highlights = merged
        .into_iter()
        .map(|range| Highlight {
            location: prefix
                + graphemes[start..range.start]
                    .iter()
                    .map(|s| s.encode_utf16().count())
                    .sum::<usize>(),
            length: graphemes[range]
                .iter()
                .map(|s| s.encode_utf16().count())
                .sum(),
        })
        .collect();
    (text, highlights)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn matching_and_offsets_preserve_original_unicode() {
        assert_eq!(normalize("CAFÉ"), normalize("Cafe\u{301}"));
        assert_eq!(normalize("Straße"), normalize("STRASSE"));
        assert_eq!(normalize("Σςσ"), "σσσ");
        let original = "👩🏽‍💻 Cafe\u{301}, CAFÉ and Straße.";
        let (text, highlights) = excerpt(original, &terms("café STRASSE"));
        assert_eq!(text, original);
        let utf16: Vec<_> = text.encode_utf16().collect();
        let matches: Vec<_> = highlights
            .iter()
            .map(|range| {
                String::from_utf16(&utf16[range.location..range.location + range.length]).unwrap()
            })
            .collect();
        assert_eq!(matches, ["Cafe\u{301}", "CAFÉ", "Straße"]);
        assert_eq!(
            excerpt("ß", &terms("s ss")).1,
            vec![Highlight {
                location: 0,
                length: 1
            }]
        );
        assert_eq!(terms(".* [x] OR"), [".*", "[x]", "or"]);
    }

    #[test]
    fn excerpts_are_bounded_without_splitting_graphemes() {
        let source = format!("{} needle {}", "👩🏽‍💻".repeat(300), "e\u{301}".repeat(300));
        let (text, highlights) = excerpt(&source, &terms("needle"));
        assert_eq!(text.graphemes(true).count(), 240);
        assert!(text.starts_with('…') && text.ends_with('…'));
        assert_eq!(highlights.len(), 1);
        let (oversized, _) = excerpt(&format!("a{}", "\u{301}".repeat(20_000)), &[]);
        assert_eq!(oversized, "…");
        assert_eq!(excerpt("", &[]), (String::new(), vec![]));
    }
}
